src/main/java/build/buf/protovalidate/internal/celext/Format.java

```diff
@@ -91,37 +83,33 @@ static Val format(String fmtString, ListT list) {
           precision = precision * 10 + (fmtString.charAt(index++) - '0');
         }
         if (index >= fmtString.length()) {
-          return Err.newErr("format: expected format specifier");
+          throw new ErrException("format: expected format specifier");
         }
         c = fmtString.charAt(index++);
       }
 
-      Val status;
       switch (c) {
         case 'd':
-          status = formatDecimal(builder, arg);
+          formatDecimal(builder, arg);
           break;
         case 'x':
-          status = formatHex(builder, arg, LOWER_HEX_ARRAY);
+          formatHex(builder, arg, LOWER_HEX_ARRAY);
           break;
         case 'X':
-          status = formatHex(builder, arg, HEX_ARRAY);
+          formatHex(builder, arg, HEX_ARRAY);
           break;
         case 's':
-          status = formatString(builder, arg);
+          formatString(builder, arg);
           break;
         case 'e':
         case 'f':
         case 'b':
         case 'o':
         default:
-          return Err.newErr("format: unparsable format specifier %s", c);
-      }
-      if (status.type() == Err.ErrType) {
-        return status;
```

Comment:
A lot of the complexity in here went away with switching to exceptions for control flow.